lib/diagnostics_channel.js

```diff
@@ -65,7 +65,7 @@ function markActive(channel) {
 
 function maybeMarkInactive(channel) {
   // When there are no more active subscribers or bound, restore to fast prototype.
-  if (!channel._subscribers.length && !channel._stores.size) {
+  if ((!channel._subscribers.length && !channel._stores.size) || channel._subscribers.every((v) => v === undefined)) {
```

Comment:
Also `[].every(v => v === undefined)` is `true`

so the `!channel._subscribers.length` check wouldn't be necessary, right?

Reply:
@iunanua Yes but the length check is probably faster no ? I'd rather execute `.every()` only when really necessary

Reply:
@szegedi actually good question, i'm not sure which one we want, i'll look into it :+1:

Reply:
@szegedi is correct, this change is needed otherwise it would be truthy if all subscribers are undefined but there _are_ still stores.

Rather than this approach of deferring removals though I think what I would actually do is to make subscribe/unsubscribe create a new array each time so the reference that has been acquired in the loop would remain constant. Rather than a push to the array you can use concat to make a new array, and for the removal you can compose a new array from a concat of the slice up to the element _before_ the position of the subscriber to remove and then concat the elements _after_ the subscriber to remove.

Reply:
@Qard wouldn't this concat approach break this pattern that currently works:
```javascript
channel.subscribe(() => {
  // oops, more stuff should be subscribed
  channel.subscribe(handlerB) // handlerB won't be called anymore

  // oops, less stuff should be subscribed
  channel.unsubscribe(handlerC) // handlerC will still be called now
})
```
not sure if that pattern is used anywhere or if it's even useful, but i'm pretty sure it would break it.
I was trying to make the API completely compatible, hence the weird nextTick().

Reply:
I made the change locally and this hypothetical test would fail:
```javascript
const onMessageHandler = common.mustCall(() => {
  dc.unsubscribe(channel_name, onMessageHandler)
  dc.subscribe(channel_name, common.mustCall()); // not called
});

dc.subscribe(channel_name, onMessageHandler);
dc.subscribe(channel_name, common.mustCall());

// This must not throw.
dc.channel(channel_name).publish(published_data);

```

Reply:
I'm fine changing the PR to your approach, but reviewers have to accept we're changing the behavior or the API here

Reply:
That's probably not a behaviour that ever should have been expected. Mutating state in one place while you're reading it in another is just never a good pattern. 😬

Reply:
I've changed the approach to make shallow copies instead of mutation, and the condition Attila was talking about is now irrelevant. I think we can close this thead.